src/Mod/Sketcher/Gui/ViewProviderSketch.cpp

```diff
@@ -3824,31 +3828,69 @@ void ViewProviderSketch::generateContextMenu()
     // if something is selected, count different elements in the current selection
     if (selection.size() > 0) {
         const std::vector<std::string> SubNames = selection[0].getSubNames();
+        const Sketcher::SketchObject* obj =
+            static_cast<Sketcher::SketchObject*>(selection[0].getObject());
 
         for (auto& name : SubNames) {
+            int geoId = std::atoi(name.substr(4, 4000).c_str()) - 1;
+            const Part::Geometry* geo = getSketchObject()->getGeometry(geoId);
             if (name.substr(0, 4) == "Edge") {
                 ++selectedEdges;
 
-                int geoId = std::atoi(name.substr(4, 4000).c_str()) - 1;
                 if (geoId >= 0) {
-                    const Part::Geometry* geo = getSketchObject()->getGeometry(geoId);
                     if (isLineSegment(*geo)) {
                         ++selectedLines;
                     }
+                    else if (geo->is<Part::GeomBSplineCurve>()) {
+                        ++selectedBsplines;
+                    }
                     else {
                         ++selectedConics;
                     }
                 }
             }
             else if (name.substr(0, 4) == "Vert") {
                 ++selectedPoints;
+                Sketcher::PointPos posId;
+                getIdsFromName(name, obj, geoId, posId);
+                if (isBsplineKnotOrEndPoint(obj, geoId, posId)) {
+                    ++selectedBsplineKnots;
+                }
             }
             else if (name.substr(0, 4) == "Cons") {
                 ++selectedConstraints;
             }
+            else if (name.substr(2, 5) == "Axis") {
+                ++selectedEdges;
+                ++selectedLines;
+                ++selectedOrigin;
+            }
+            else if (name.substr(0, 4) == "Root") {
+                ++selectedPoints;
+                ++selectedOrigin;
+            }
+        }
+        if (selectedPoints + selectedEdges == selectedOrigin) {
+            onlyOrigin = true;
         }
         // build context menu items depending on the selection
-        if (selectedEdges >= 1 && selectedPoints == 0) {
+        if (selectedBsplines > 0 && selectedBsplines == selectedEdges && selectedPoints == 0
+            && !onlyOrigin) {
+            menu << "Sketcher_BSplineInsertKnot"
+                 << "Sketcher_BSplineIncreaseDegree"
+                 << "Sketcher_BSplineDecreaseDegree";
+        }
+        else if (selectedBsplineKnots > 0 && selectedBsplineKnots == selectedPoints
+                 && selectedEdges == 0 && !onlyOrigin) {
+            if (selectedBsplineKnots == 1) {
+                menu << "Sketcher_BSplineIncreaseKnotMultiplicity"
+                     << "Sketcher_BSplineDecreaseKnotMultiplicity";
+            }
+            else if (selectedBsplineKnots == 2) {
+                menu << "Sketcher_JoinCurves";
```

Comment:
Join curves works for any two curves, not just bsplines. And only end points should be selected, not knots.

Reply:
you are right, but it converts the "normal" geometry to splines. Not sure if we should add that in the context menu if not a spline end point is selected.

Reply:
Why not? If we have selected two separate end points it does seem to make sense.

Reply:
true. added it.